AWS DescribeAutoScalingGroups requests too aggressive - API limits reached

mmerrill3 opened the pull request:
Fix for issue 252 where excessive API calls to AWS are the result of leaking go routines from the Polling auto scaler.

mmerrill3 commented:
on a side note, I have a mac, and it seems even if I set my GOOS to linux, I am having issues building/testing locally using the master as my baseline.

See below:

mmerrillmbp:cluster-autoscaler mmerrill$ go version
go version go1.8.3 darwin/amd64
mmerrillmbp:cluster-autoscaler mmerrill$ 
mmerrillmbp:cluster-autoscaler mmerrill$ 
mmerrillmbp:cluster-autoscaler mmerrill$ 
mmerrillmbp:cluster-autoscaler mmerrill$ 
mmerrillmbp:cluster-autoscaler mmerrill$ make test-unit
rm -f cluster-autoscaler
go get github.com/tools/godep
GOOS=linux godep go build ./...
vendor/github.com/google/cadvisor/container/crio/factory.go:24:2: no buildable Go source files in /Users/mmerrill/go/src/k8s.io/autoscaler/cluster-autoscaler/vendor/github.com/google/cadvisor/container/libcontainer
godep: go exit status 1
make: *** [build] Error 1

MaciekPytel commented:
@mmerrill3 Can you squash your commits and add a comment once this is ready for review? Thanks.

mmerrill3 commented:
@MaciekPytel, the commits are squashed and this is ready for a review.

mumoshu commented on cluster-autoscaler/cloudprovider/cloud_provider.go:
```diff
@@ -57,6 +57,9 @@ type CloudProvider interface {
 	GetResourceLimiter() (*ResourceLimiter, error)
 
+	// Close cleans up open resources before the cloud provider is destroyed, i.e. go routines etc.
+	Close() error
```
nit: In case there's no difference, could we rename this `Close` to `Cleanup` so that there are less func names to remember? 😉

mmerrill3 replied:
ok, will update this with the renamed interface method soon.

mumoshu replied:
@mmerrill3 Thanks for catching it!

I'd appreciate it if you could also rebase this onto the latest master so that hopefully @MaciekPytel can merge this earlier before this gets conflicted again 😃

mmerrill3 replied:
I am handling rebasing now.

mumoshu commented on cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go:
```diff
@@ -40,15 +38,6 @@ func newAutoScalingGroups(service autoScalingWrapper) *autoScalingGroups {
-		if err := registry.regenerateCache(); err != nil {
-			glog.Errorf("Error while regenerating Asg cache: %v", err)
-		}
-	}, time.Hour)
```
Ah, this does look bad. Thanks for the good catch!

mumoshu commented:
One nit, but LGTM. Thank you very much for trying to fix the issue 👍

mwielgus commented:
@mmerrill3 my attempt to resolve conflicts in github failed. Could you please do it yourself and push the new commit?

mwielgus commented:
```
 k8s.io/autoscaler/cluster-autoscaler/cloudprovider/gce
cloudprovider/gce/gce_cloud_provider.go:169: ambiguous selector gce.gceManager.GetResourceLimiter
cloudprovider/gce/gce_manager.go:109: duplicate method GetResourceLimiter
# k8s.io/autoscaler/cluster-autoscaler/cloudprovider/gce
cloudprovider/gce/gce_cloud_provider.go:169: ambiguous selector gce.gceManager.GetResourceLimiter
cloudprovider/gce/gce_manager.go:109: duplicate method GetResourceLimiter
godep: go exit status 2
```

And please squash the commits.

mmerrill3 commented:
@mwielgus, thanks for your patience, I have rebased my forked master branch, squashed the commits, and updated the interface method per feedback from @mumoshu.

mwielgus commented:
/lgtm
Thank you for fixing this! :)

johanneswuerbach commented:
Any chance to get this cherry picked into the next 1.0 patch release? Would this require a cherry pick PR?

mwielgus commented:
@johanneswuerbach - please create a cherry-pick

johanneswuerbach commented:
@mwielgus done https://github.com/kubernetes/autoscaler/pull/445
